packages/one-app-dev-bundler/esbuild/plugins/externals-loader.js

```diff
@@ -40,20 +55,36 @@ const externalsLoader = ({ bundleType }) => ({
     // return a namespace. (see this above)
     // your onLoad can then just match .* within that namespace and you guarantee you target
     // every package you want.
-    build.onLoad({ filter: /.*/, namespace: 'externalsLoader' }, async (args) => {
-      const jsContent = `\
-try {
-  module.exports = ${globalReferenceString}.getTenantRootModule().appConfig.providedExternals['${args.path}'].module;
-} catch (error) {
-  const errorGettingExternal = new Error('Failed to get external ${args.path} from root module');
-  errorGettingExternal.shouldBlockModuleReload = false;
-  throw errorGettingExternal;
-}
-`;
+    build.onLoad({ filter: /.*/, namespace: 'externalsLoader' }, async ({ path: externalName }) => {
+      const version = readPackageUpSync({
+        cwd: path.resolve(process.cwd(), 'node_modules', externalName),
+      })?.packageJson.version;
 
       return {
-        contents: jsContent,
         loader: 'js',
+        contents: `
+          try {
+            const Holocron = ${bundleType === BUNDLE_TYPES.SERVER ? 'require("holocron")' : `${globalReferenceString}.Holocron`};
+            const fallbackExternal = Holocron.getExternal({
+              name: '${externalName}',
+              version: '${version}'
+            });
+            const rootModuleExternal = ${globalReferenceString}.getTenantRootModule && ${globalReferenceString}.getTenantRootModule().appConfig.providedExternals['${externalName}'];
+
+            module.exports = fallbackExternal || (rootModuleExternal ? rootModuleExternal.module : () => {
```

Comment:
If there is both a fallback external, and a root module external, why would we put the fallback external in.

If the root module external exists, should we not use that by priority?

Reply:
One App and Holocron are responsible for only adding fallbacks when required. If the root module provided a valid version there should not be a fallback.